crates/starknet/src/starknet/state_update.rs

```diff
@@ -0,0 +1,147 @@
+    use crate::state::state_diff::StateDiff;
+    use crate::state::state_update::StateUpdate;
+    use crate::traits::{Accounted, HashIdentifiedMut};
+    use crate::transactions::declare_transaction_v2::DeclareTransactionV2;
+    use crate::utils::load_cairo_0_contract_class;
+    use crate::utils::test_utils::{dummy_cairo_1_contract_class, dummy_felt};
+
+    #[test]
+    /// This test checks that the state update is correct after a declare transaction v2.
+    /// Then checks that the state update is empty after executing the same declare transaction
+    fn correct_state_update_after_declare_transaction_v2() {
+        let (mut starknet, sender_address) = setup();
+        let contract_class = dummy_cairo_1_contract_class();
+
+        let sierra_class_hash = contract_class.generate_hash().unwrap();
+        let casm_contract_class = CasmContractClass::try_from(contract_class.clone()).unwrap();
+        let compiled_class_hash = compute_casm_class_hash(&casm_contract_class).unwrap();
+
+        let declare_txn = DeclareTransactionV2::new(
+            contract_class,
+            compiled_class_hash.clone().into(),
+            sender_address,
+            100,
+            Vec::new(),
+            Felt::from(0),
+            StarknetChainId::TestNet.to_felt().into(),
+        )
+        .unwrap();
+
+        // first execute declare v2 transaction
+        let (txn_hash, _) = starknet.add_declare_transaction_v2(declare_txn.clone()).unwrap();
+        assert_eq!(
+            starknet.transactions.get_by_hash_mut(&txn_hash).unwrap().status,
+            TransactionStatus::AcceptedOnL2
+        );
+        let state_update = starknet
+            .block_state_update(starknet_rs_core::types::BlockId::Tag(
+                starknet_rs_core::types::BlockTag::Latest,
+            ))
+            .unwrap();
+
+        let expected_state_diff = StateDiff {
+            declared_contracts: vec![(compiled_class_hash.clone().into(), casm_contract_class)]
+                .into_iter()
+                .collect(),
+            class_hash_to_compiled_class_hash: vec![(
+                sierra_class_hash,
+                compiled_class_hash.into(),
+            )]
+            .into_iter()
+            .collect(),
+            ..Default::default()
+        };
+
+        let expected_state_update = StateUpdate::new(Felt::default(), expected_state_diff).unwrap();
+
+        // check only 3 of the 4 fields, because the inner property has changes to the storage of
+        // the ERC20 contract which are hard to be tested correctly, it depends on the fee
+        // calculation of starknet_in_rust_library
+        assert_eq!(
+            state_update.cairo_0_declared_classes,
+            expected_state_update.cairo_0_declared_classes
+        );
+        assert_eq!(state_update.declared_classes, expected_state_update.declared_classes);
+
+        let (txn_hash, _) = starknet.add_declare_transaction_v2(declare_txn).unwrap();
```

Comment:
[c9c7a80](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/103/commits/c9c7a803ca6a60bf8460230a31a469f54c6f640f)
The fact that its not returning rejected transaction is because the account_without_validations is used